### PR TITLE
Reader: only show unique authors with avatars in conversation caterpillar

### PR DESCRIPTION
**Pull request by bluefuton**
In the Reader conversation caterpillar:
- [x] show only unique author avatars
- [x] omit authors without an avatar URL
- [x] count should reflect ~unique authors~ comment count
- [x] last author name should reflect last displayed avatar

screen shot 2017-08-03 at 18 24 38

### To test

http://calypso.localhost:3000/devdocs/blocks/conversation-caterpillar

### CODE REVIEW

**Comment by fraying** [comment]
Note: The number is the number of comments, not authors.

**Comment by bluefuton** [comment]
> Note: The number should be the number of comments, not authors.

@fraying I wondered about how that'd feel if one author dominates the discussion.

For example, I might have 7 comments and I've written 6 of them. It might say:

( avatar1 ) ( avatar2 ) Chris and 6 more

I think I'd probably read it as "Chris and 6 other people contributed to the discussion" rather than "Chris R commented and there are 6 other comments besides that".

What do you think @blowery + @samouri?

**Comment by fraying** [comment]
@bluefuton We may play with the exact text to make it more clear. But it's the number of comments we need to communicate.

File: client/blocks/conversation-caterpillar/index.jsx [hunk 1]
```diff
@@ -21,19 +22,22 @@ class ConversationCaterpillarComponent extends React.Component {
...
 	render() {
 		const { comments, translate } = this.props;
-		const lastComment = last( comments );
-		const lastCommenterName = get( lastComment, 'author.name' );
-		const commentCount = comments.length;
+		const commentCount = size( comments );
+
+		// Only display authors with a gravatar, and only display each author once
+		const uniqueAuthors = uniqBy( map( comments, 'author' ), 'ID' );
+		const displayedAuthors = filter( uniqueAuthors, author => !! author.avatar_URL );
```
Comment (samouri) [inline client/blocks/conversation-caterpillar/index.jsx]: nittiest nit and not even better, but this is cool: `const displayAuthors = reject( uniqueAuthors, 'avatar_URL' );`

File: client/blocks/conversation-caterpillar/index.jsx [hunk 2]
```diff
@@ -21,19 +22,22 @@ class ConversationCaterpillarComponent extends React.Component {
 
 	render() {
 		const { comments, translate } = this.props;
-		const lastComment = last( comments );
-		const lastCommenterName = get( lastComment, 'author.name' );
-		const commentCount = comments.length;
+		const commentCount = size( comments );
```
Comment (samouri) [inline client/blocks/conversation-caterpillar/index.jsx]: do we want to show actual comment count or just ones available to the :bug:?
I would think you actually want to use `post.discussion.comment_count`

File: client/blocks/conversation-caterpillar/index.jsx (hunk 1 above)
Comment (bluefuton) [inline client/blocks/conversation-caterpillar/index.jsx]: I like that!

Comment (bluefuton) [inline client/blocks/conversation-caterpillar/index.jsx]: Ah - won't `reject` give us the elements for which `author.avatar_URL` is falsey? I suspect we do want `filter` after all.

I'll use the property shorthand though!

File: client/blocks/conversation-caterpillar/index.jsx (hunk 2 above)
Comment (bluefuton) [inline client/blocks/conversation-caterpillar/index.jsx]: @fraying explained in Slack:

> I was assuming the number was all comments for that post, minus however many are expanded already.
So, a post with 100 comments total, with three expanded in Convos, would be 97
A post with 100 comments total, with 10 expanded on the Fullpost page, the number would be 90
In Convos, if you closed all the comments, that 100 comment post would show 100.

Sounds like we'll have to pass the count into the component eventually from a parent component. I'll leave this as it is for the moment.

**Pull request merged**